include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -2477,47 +3160,1289 @@ namespace internal
-      const MatrixFreeFunctions::ShapeInfo<Number> &shape_info,
-      const Number *                                input,
+    evaluate_in_face(const EvaluationFlags::EvaluationFlags evaluation_flag,
+                     const MatrixFreeFunctions::ShapeInfo<Number> &shape_info,
+                     Number *                                      values_dofs,
+                     Number *                                      values_quad,
+                     Number *           gradients_quad,
+                     Number *           hessians_quad,
+                     Number *           scratch_data,
+                     const unsigned int subface_index,
+                     const unsigned int face_no)
+    {
+      // TODO. Need both normal/tangent at 0 and 1 since it depends on the
+      // component!
+
+      Eval_normal eval_normal =
+        create_evaluator_tensor_product<Eval_normal>(shape_info.data.front(),
+                                                     subface_index,
+                                                     0);
+      Eval_tangent eval_tangent =
+        create_evaluator_tensor_product<Eval_tangent>(shape_info.data.back(),
+                                                      subface_index,
+                                                      1);
+
+      // Used for normal faces which are not anisotropic
+      Eval_general eval_general(shape_info.data.back().shape_values,
+                                shape_info.data.back().shape_gradients,
+                                shape_info.data.back().shape_hessians,
+                                shape_info.data.back().fe_degree + 1,
+                                shape_info.data.back().n_q_points_1d);
+
+      // Note, n_dofs on tangent face
+      const std::size_t n_dofs_tangent = shape_info.dofs_per_component_on_face;
+      const std::size_t n_dofs_normal =
+        n_dofs_tangent - Utilities::pow(fe_degree, dim - 2);
+
+      const std::size_t n_q_points =
+        fe_degree > -1 ?
+          Utilities::pow(n_q_points_1d, dim - 1) :
+          Utilities::pow(shape_info.data.front().n_q_points_1d, dim - 1);
+
+      const unsigned int face_orientation = face_no / 2;
+
+      // This can probably be implemented better.
+      // TODO. Here one needs to check subindex and choose direction.
+      if (face_orientation == 0)
+        {
+          evaluate_in_normal_face_apply<Eval_general, Eval_general>(
+            eval_general,
+            eval_general,
+            values_dofs,
+            values_quad,
+            gradients_quad,
+            hessians_quad,
+            scratch_data,
+            evaluation_flag,
+            n_dofs_normal,
+            n_q_points);
+
+          values_dofs += 3 * n_dofs_normal;
+          values_quad += n_q_points;
+          gradients_quad += dim * n_q_points;
+          hessians_quad += dim * (dim + 1) / 2 * n_q_points;
+
+
+
+          evaluate_in_tangent_face_apply<Eval_normal, Eval_tangent, 0>(
```

Comment:
If you put the `0` template argument first, can the other two template arguments be picked up automatically through the function arguments?